### PR TITLE
Mbarba/prepare tweaks

### PR DESCRIPTION
**Pull request by MatBarba**
To show comparable stats to what's in RefSeq assembly pages

### CODE REVIEW

File: lib/python/ensembl/brc4/runnable/manifest_stats.py [hunk 1]
```diff
@@ -143,25 +143,40 @@ def parse_gff3(self, gff3_handle: TextIO) -> List:
...
+                if feat1.type in ("gene", "pseudogene"):
+                    self.increment_biotype(biotypes, feat1.id, "ALL_GENES")
                 
         # Order
         sorted_biotypes = OrderedDict()
         for name in sorted(biotypes.keys()):
-            sorted_biotypes[name] = biotypes[name]
+            data = biotypes[name]
+            counter = Counter(data["ids"])
+            data["unique_count"] = len(counter.keys())
```
Comment (JAlvarezJarreta) [inline lib/python/ensembl/brc4/runnable/manifest_stats.py:169]: An alternative (would be even shorter if `data["ids"]` is already a set (see my next comment)):
```suggestion
            data["unique_count"] = len(set(data["ids"]))
```

File: lib/python/ensembl/brc4/runnable/manifest_stats.py [hunk 2]
```diff
@@ -143,25 +143,40 @@ def parse_gff3(self, gff3_handle: TextIO) -> List:
...
+        stats = [f"{data['unique_count']:>9}\t{biotype:<20}\tID = {data['example']}" for (biotype, data) in sorted_biotypes.items()]
         
         return stats
 
-    def increment_biotype(self, biotypes: Dict, feature) -> None:
-        biotype = feature.type
-        if not biotype in biotypes:
-            biotypes[biotype] = { "count" : 0, "example" : feature.id }
-        biotypes[biotype]["count"] += 1
+    def increment_biotype(self, biotypes: Dict, feature_id: str, feature_biotype: str) -> None:
```
Comment (JAlvarezJarreta) [inline lib/python/ensembl/brc4/runnable/manifest_stats.py:176]: This method seems to be related to this class, but does not need to know anything about it:
```suggestion
    @staticmethod
    def increment_biotype(biotypes: Dict, feature_id: str, feature_biotype: str) -> None:
```
If you ahead with this change, all the previous call should be replaced from `self.increment_biotype()` to `manifest_stats.increment_biotype()`.

**Review by JAlvarezJarreta — APPROVED** [review]
Just a few comment with some suggestions, but the code is good as it is :+1:

File: lib/python/ensembl/brc4/runnable/process_gff3.py [hunk 3]
```diff
@@ -636,42 +636,110 @@ def build_transcript(self, gene: SeqFeature) -> SeqFeature:
...
+                cdss.append(subf)
+            else:
+                gene_subf_clean.append(subf)
+
+        if len(cdss) > 0:
+            if len(mrnas) > 1:
+                raise Exception(f"Can't fix gene {gene.id}: contains several mRNAs and CDSs, all children of the gene")
+        else:
+            # Nothing to fix here, no CDSs to move
+            return gene
```
Comment (JAlvarezJarreta) [inline lib/python/ensembl/brc4/runnable/process_gff3.py:674]: Suggestion with less lines:
```suggestion
        if len(cdss) == 0:
            # Nothing to fix here, no CDSs to move
            return gene
        if len(mrnas) > 1:
            raise Exception(f"Can't fix gene {gene.id}: contains several mRNAs and CDSs, all children of the gene")
```

File: lib/python/ensembl/brc4/runnable/process_gff3.py [hunk 4]
```diff
@@ -636,42 +636,110 @@ def build_transcript(self, gene: SeqFeature) -> SeqFeature:
...
 
     def move_cds_to_mrna(self, gene: SeqFeature) -> SeqFeature:
-        """Move a CDS child of a gene, to the mRNA.
+        """Move CDS child features of a gene, to the mRNA.
+
+        This is to fix the case where we have the following structure:
+        gene -> [ mRNA, CDSs ]
+        and change it to 
+        gene -> [ mNRA -> [ CDSs ] ]
+        The mRNA might have exons, in which case check that they match the CDS coordinates
```
Comment (JAlvarezJarreta) [inline lib/python/ensembl/brc4/runnable/process_gff3.py:645]:
```suggestion
        The mRNA might have exons, in which case check that they match the CDS coordinates.
```

File: lib/python/ensembl/brc4/runnable/process_gff3.py [hunk 5]
```diff
@@ -636,42 +636,110 @@ def build_transcript(self, gene: SeqFeature) -> SeqFeature:
...
+        gene.sub_features = gene_subf_clean
+        gene.sub_features.append(mrna)
         
-        return mrna
+        return gene
+    
+    def clean_extra_exons(self, gene: SeqFeature) -> SeqFeature:
+        """Remove extra exons, already existing in the mRNA.
+
+        This is a special case where a gene contains proper mRNAs etc. but also
```
Comment (JAlvarezJarreta) [inline lib/python/ensembl/brc4/runnable/process_gff3.py:713]:
```suggestion
        This is a special case where a gene contains proper mRNAs, etc. but also
```

File: lib/python/ensembl/brc4/runnable/process_gff3.py [hunk 6]
```diff
@@ -636,42 +636,110 @@ def build_transcript(self, gene: SeqFeature) -> SeqFeature:
...
+            if subf.type == "exon":
+                exons.append(subf)
+            elif subf.type == "mRNA":
+                mrnas.append(subf)
+            else:
+                others.append(subf)
+        
+        if exons and mrnas:
+            exon_has_id = 0
+            # Check if the exon ids start with id-, which is an indication that they do not belong here
```
Comment (JAlvarezJarreta) [inline lib/python/ensembl/brc4/runnable/process_gff3.py:730]:
```suggestion
            # Check if the exon ids start with "id-", which is an indication that they do not belong here
```

File: lib/python/ensembl/brc4/runnable/process_gff3.py [hunk 7]
```diff
@@ -636,42 +636,110 @@ def build_transcript(self, gene: SeqFeature) -> SeqFeature:
...
+            for exon in exons:
+                if exon.id.startswith("id-"):
+                    exon_has_id += 1
+            if exon_has_id:
+                if exon_has_id == len(exons):
+                    print(f"Remove {exon_has_id} extra exons from {gene.id}")
+                    gene.sub_features = mrnas
+                    gene.sub_features += others
+                else:
+                    raise Exception(f"Can't remove extra exons for {gene.id}, some have gene id-, others do not")
```
Comment (JAlvarezJarreta) [inline lib/python/ensembl/brc4/runnable/process_gff3.py:740]:
```suggestion
                    raise Exception(f"Can't remove extra exons for {gene.id}, not all start with 'id-'")
```
(to fit it in one single line but adding `'` as needed)

**Review by JAlvarezJarreta — CHANGES_REQUESTED** [review]

File: lib/python/ensembl/brc4/runnable/manifest_stats.py (hunk 2 above)
Comment (MatBarba) [reply lib/python/ensembl/brc4/runnable/manifest_stats.py]: Done

File: lib/python/ensembl/brc4/runnable/manifest_stats.py [hunk 8]
```diff
@@ -143,25 +143,40 @@ def parse_gff3(self, gff3_handle: TextIO) -> List:
...
+        if not feature_biotype in biotypes:
+            biotypes[feature_biotype] = { "count" : 0, "ids": [], "example" : feature_id }
+        biotypes[feature_biotype]["count"] += 1
+        # Only count IDs once!
+        biotypes[feature_biotype]["ids"].append(feature_id)
```
[the remark this answers is not in the document]
Comment (MatBarba) [reply lib/python/ensembl/brc4/runnable/manifest_stats.py]: It is indeed much simpler with a set

File: lib/python/ensembl/brc4/runnable/manifest_stats.py (hunk 1 above)
Comment (MatBarba) [reply lib/python/ensembl/brc4/runnable/manifest_stats.py]: Now using a set

File: lib/python/ensembl/brc4/runnable/manifest_stats.py [hunk 9]
```diff
@@ -15,7 +15,7 @@
 # limitations under the License.
 
 
-from collections import OrderedDict
+from collections import Counter, OrderedDict
```
[the remark this answers is not in the document]
Comment (MatBarba) [reply lib/python/ensembl/brc4/runnable/manifest_stats.py]: Removed, thanks

**Review by JAlvarezJarreta — APPROVED** [review]
Good to go!

**Pull request merged**